File: java/src/main/java/it/giannibombelli/wsc2026/payment/domain/payment/Transaction.java
package it.giannibombelli.wsc2026.payment.domain.payment;

import it.giannibombelli.wsc2026.common.utils.Require;

import it.giannibombelli.wsc2026.common.domain.model.Entity;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.common.domain.primitive.Timestamp;

import java.util.UUID;

public final class Transaction implements Entity<TransactionId> {
    private final TransactionId id;
    private final Provider provider;
    private final ProviderReference providerReference;
    private final Money amount;
    private final TransactionStatus status;
    private final Timestamp startedAt;
    private final Timestamp completedAt;

    public Transaction(TransactionId id, Provider provider, ProviderReference providerReference, Money amount,
                       TransactionStatus status, Timestamp startedAt, Timestamp completedAt) {
        Require.requireArgument(id, "id");
        Require.requireArgument(provider, "provider");
        Require.requireArgument(amount, "amount");
        Require.requireArgument(status, "status");
        Require.requireArgument(startedAt, "startedAt");
        this.id = id;
        this.provider = provider;
        this.providerReference = providerReference;
        this.amount = amount;
        this.status = status;
        this.startedAt = startedAt;
        this.completedAt = completedAt;
    }

    public static Transaction start(TransactionId id, Provider provider, ProviderReference providerReference, Money amount, Timestamp startedAt) {
        return new Transaction(id, provider, providerReference, amount, TransactionStatus.STARTED, startedAt, null);
    }

    public Transaction accept(Timestamp completedAt) {
        if (status != TransactionStatus.STARTED) throw new IllegalStateException("transaction not started");
        Require.requireArgument(completedAt, "completedAt");
        return new Transaction(id, provider, providerReference, amount, TransactionStatus.ACCEPTED, startedAt, completedAt);
    }

    public Transaction reject(Timestamp completedAt) {
        if (status != TransactionStatus.STARTED) throw new IllegalStateException("transaction not started");
        Require.requireArgument(completedAt, "completedAt");
        return new Transaction(id, provider, providerReference, amount, TransactionStatus.REJECTED, startedAt, completedAt);
    }

    public Transaction refund(Timestamp refundedAt) {
        if (status != TransactionStatus.ACCEPTED) throw new IllegalStateException("transaction not accepted");
        Require.requireArgument(refundedAt, "refundedAt");
        return new Transaction(id, provider, providerReference, amount, TransactionStatus.REFUNDED, startedAt, refundedAt);
    }

    public TransactionId id() {
        return id;
    }

    public Provider provider() {
        return provider;
    }

    public ProviderReference providerReference() {
        return providerReference;
    }

    public Money amount() {
        return amount;
    }

    public TransactionStatus status() {
        return status;
    }

    public Timestamp startedAt() {
        return startedAt;
    }

    public Timestamp completedAt() {
        return completedAt;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        return id.equals(((Transaction) o).id);
    }

    @Override
    public int hashCode() {
        return id.hashCode();
    }
}

File: typescript/src/payment/domain/payment/transaction.ts
// Value Object/Entity Transaction all'interno dell'aggregato Payment.

import { Money } from '@/common/domain/primitive/money.js';
import { Timestamp } from '@/common/domain/primitive/timestamp.js';
import { Provider } from './provider.js';
import { ProviderReference } from './providerReference.js';
import { TransactionId } from './transactionId.js';
import { TransactionStatus } from './transactionStatus.js';
import { requireArgument } from '@/common/utils/requireArgument.js';

export class Transaction {
  private readonly _id: TransactionId;
  private readonly _provider: Provider;
  private readonly _providerReference: ProviderReference | null;
  private readonly _amount: Money;
  private readonly _status: TransactionStatus;
  private readonly _startedAt: Timestamp;
  private readonly _completedAt: Timestamp | null;

  constructor(
    id: TransactionId,
    provider: Provider,
    providerReference: ProviderReference | null,
    amount: Money,
    status: TransactionStatus,
    startedAt: Timestamp,
    completedAt: Timestamp | null
  ) {
    requireArgument(id, 'id');
    requireArgument(provider, 'provider');
    requireArgument(amount, 'amount');
    requireArgument(status, 'status');
    requireArgument(startedAt, 'startedAt');

    this._id = id;
    this._provider = provider;
    this._providerReference = providerReference;
    this._amount = amount;
    this._status = status;
    this._startedAt = startedAt;
    this._completedAt = completedAt;
  }

  static start(
    id: TransactionId,
    provider: Provider,
    providerReference: ProviderReference | null,
    amount: Money,
    startedAt: Timestamp
  ): Transaction {
    return new Transaction(id, provider, providerReference, amount, TransactionStatus.STARTED, startedAt, null);
  }

  accept(completedAt: Timestamp): Transaction {
    if (this._status !== TransactionStatus.STARTED) throw new Error('transaction not started');
    requireArgument(completedAt, 'completedAt');
    return new Transaction(
      this._id,
      this._provider,
      this._providerReference,
      this._amount,
      TransactionStatus.ACCEPTED,
      this._startedAt,
      completedAt
    );
  }

  reject(completedAt: Timestamp): Transaction {
    if (this._status !== TransactionStatus.STARTED) throw new Error('transaction not started');
    requireArgument(completedAt, 'completedAt');
    return new Transaction(
      this._id,
      this._provider,
      this._providerReference,
      this._amount,
      TransactionStatus.REJECTED,
      this._startedAt,
      completedAt
    );
  }

  refund(refundedAt: Timestamp): Transaction {
    if (this._status !== TransactionStatus.ACCEPTED) throw new Error('transaction not accepted');
    requireArgument(refundedAt, 'refundedAt');
    return new Transaction(
      this._id,
      this._provider,
      this._providerReference,
      this._amount,
      TransactionStatus.REFUNDED,
      this._startedAt,
      refundedAt
    );
  }

  id(): TransactionId {
    return this._id;
  }

  provider(): Provider {
    return this._provider;
  }

  providerReference(): ProviderReference | null {
    return this._providerReference;
  }

  amount(): Money {
    return this._amount;
  }

  status(): TransactionStatus {
    return this._status;
  }

  startedAt(): Timestamp {
    return this._startedAt;
  }

  completedAt(): Timestamp | null {
    return this._completedAt;
  }

  equals(other: Transaction): boolean {
    return this._id.equals(other._id);
  }

  hashCode(): number {
    return this._id.hashCode();
  }
}
